### ISSUE TITLE
P4_14 (v1.0.4): Order of update/verify for calculated_fields

### ISSUE THREAD

**Issue by kheradmand** [p4lang/p4-spec#442]
The specification does not specify the order of performing update/verify in case multiple calculated_fields exist. 
The order however, can affect the outcome of at least update operation by affecting the updates' guard conditions or in cases that a calculated_field's field_list contains another calculated_field. It can potentially affect the outcome of verify as well (by affecting the `parser_error_location` field in standard_metadata)

**Comment by kheradmand** [comment]
At least for the update case, a sensible order is the (reverse) order in which the headers are deparsed (so that an encapsulating header's checksum is updated after the checksum of encapsulated headers are updated).

**Comment by vgurevich** [comment]
Hello @kheradmand ,

I do agree that if a field_list for one of the calculated fields contains another calculated field, the spec is ambiguous. However, in reality this is not normally the case. In fact, I haven't seen any other use for the calculated fields than IPv4/TCP/UDP checksum calculation and all these are OK, since IPv4 checksum doesn't contain TCP/UDP checksum field and TCP (or UDP) checksum field doesn't contain IPv4 checksum field.

Given that in most HW implementations I'd expect these calculations to be somewhat parallelized, the result for the case where one calculated field references another would probably be undefined. 

Do you have a specific example of a protocol that requires this?

Thanks,
Vladimir

**Comment by jnfoster** [comment]
Ali has been doing heroic work to formalize P4_14 in the K framework -- an effort that is uncovering lots of corner cases, and which I very much appreciate it.

Also, while we do want to design the language so current protocols can be cleanly modeled, the lack of a protocol that uses any particular combination of language features is not a good reason to leave their semantics undefined. Arguably the whole point of defining a programming language rather than a library of fixed functions is to enable people to define new protocols. It's fine if we say that the semantics is non-deterministic (so implementations can do this calculation in either order) but it should not undefined -- eventually I hope we get to a state where the language specification assigns a meaning to every valid program.

**Comment by vgurevich** [comment]
Hi Ali,

Please, do not misconstrue my previous comment as dismissive of your findings. As I said in the first sentence, I totally agree with them. The question is what to do next.

Generally, I think we should clarify the semantics or the problems where we can. In this case, I think it would be a very good idea to add a phrase that would describe what happens if you try to have a calculated field whose calculation refers to another calculated. field. Personally, I think this is something that should not be allowed and P4 compiler can easily check for that. 

That was the reason for my question about whether you have a specific need for this.  If (as I believe) there is no practical need to have calculated fields calculated over other calculated fields,  then if we disallow this, nobody will hurt and the language becomes better defined at the same time. 

The other option is to say that the behavior is implementation-specific. Again, this will basically have the effect of saying "don't do this" and therefore I'd like to do it only while being reasonably certain that we are not going to exclude legitimate programs. 

The third option, to define a specific sequential behavior as you proposed, is also possible, but I'd employ it only if I am reasonably sure that there are a number of programs that people want to write, but can't unless this behavior is clarified. The reason I say this is because mandating such a behavior will impose unnecessary burden on a number of implementations, specifically on the high-speed HW that tends to do such things in parallel, so it would be a pity if we mandate this behavior, but then nobody will write any interesting programs that rely on it.

Hope, this explains my approach better.

Thanks,
Vladimir

**Comment by jnfoster** [comment]
I totally agree. Specifying sequential order seems too strong. Let's either outlaw this or specify that the behavior is non-deterministic.

**Comment by kheradmand** [comment]
Hi Vladimir, Hi Nate,

Thanks a lot for your responses :) 
As Nate mentioned (also thanks for flattering comment on my work) my question was purely from perspective of language semantics. I actually wasn't thinking about any practical use cases.
My comment regarding a specific order was based on my vague intuition about layering in networking and really not an expert suggestion. Actually in my K definition, I use a non-deterministic order to support the most general behavior (so that one can search all possible behaviors and see if any of them leads to any problems)

Based on your discussion, I think only mentioning the order is implementation-specific (or non-deterministic) (1) or only disallowing calculated field from referencing each other (or use of a calculated field in the guard condition of an update/verify) (2) is not enough, and both of this two things should be mentioned in the spec, because
- only (1): As you said if a target does the calculations in parallel, then the is no sequential order and the semantics of calculated fields that refer to other calculated fields is not clear. 
- only (2): still it is not clear what happens since the order can potentially affect the outcome of verify by affecting the parser_error_location field in standard_metadata.

**Comment by kheradmand** [comment]
@vgurevich 

> Generally, I think we should clarify the semantics or the problems where we can. In this case, I think it would be a very good idea to add a phrase that would describe what happens if you try to have a calculated field whose calculation refers to another calculated. field. Personally, I think this is something that should not be allowed **and P4 compiler can easily check for that.**

I am not sure if this can be determined at compile time if a field_list contains `payload` keyword.

**Comment by vgurevich** [comment]
@kheradmand ,

I am not sure I understand the relationship between the calculated fields calculated over other calculated fields and payload. Payload doesn't change to begin with.

**Comment by kheradmand** [comment]
@vgurevich,

Maybe the problem is my understanding of the semantics of the `payload` keyword in field_list (I have opened [another issue](https://github.com/p4lang/p4-spec/issues/433) regarding this). 

The spec says: 

> The identifier payload indicates that the contents of the packet following the header of the previously mentioned field is included in the field list.

I don't know what "previously mentioned field" exactly means but I guess in something line 
```
field_list {
   h.f;
   payload;
}
```
It probably means the content of the packet following h. If this reading is correct, the "the content of the packet following h" might include another header g for example which might contain a calculated field.

**Comment by vgurevich** [comment]
@kheradmand ,

I see... Your understanding of "payload" is correct. 

We can mandate that any changes to the payload must be finalized before the checksum is updated.

In practice, though, high-speed devices to not calculate the checksum of the payload. Instead, they derive it from the calculated field in the parser by subtracting the values of all the other fields from the calculated field. Then, this residual checksum (which is effectively the checksum of the payload) is carried through the match-action pipeline and then in the deparser the values of all other fields are added back to the checksum.  If someone changes the payload (as you understand it) in the match-action pipeline, they are basically screwed.
